### DISCUSSION TITLE
Dependency on SQLite?

### DISCUSSION THREAD

**Discussion by bhw-foss** [rabbibotton/clog#112] [Q&A] [answered]
Good morning Dr. Botton,

First things first, thank you for CLOG. Especially the tutorials and great documentation. For a portable, lispier persistence solution concerning my hobby project I use [Rucksack](https://gitlab.common-lisp.net/rucksack/rucksack/-/blob/master/doc/talk-eclm2006.txt) (embedded, open source version of Allegrocache) in lieu of SQLite. I also desire to remain a clogger! Are there any plans to let me have my cake and eat it too? Right now I've just included sqlite as part of my build dependencies. 

I know you've `:depends-on` CL-DBI, but maybe CLSQL `.asd` files might be useful as an example? I know CLSQL manages multiple backends.

Cheers,

Hanshen

**Comment by rabbibotton** [comment, marked as answer]
Up until now I have only used sqlserver and cl-dbi for some demos. Up until
now you really don't need any databases. I am in the middle of adding data
bindings and was going to use the dbi interface for the database access
since it had a few backends, but not there yet, so your timing is perfect :)

I will for sure look into rucksack if adding any source of db use to CLOG
itself as I want it pure lisp as much as possible.

I assumed since any reasonable system today has sqlserver as part of the OS
(window is not reasonable but easy enough to add the dll) so would not be
harmful to require it for CLOG.

Do you see an issue with the inclusion of sqlserver an issue?
I could also can move all the database dependencies to their own asdf
system (say :clog/db) would that solve things for you?
Have you considered adding a cl-dbi driver for rucksack?

**Reply by bhw-foss** [reply]
Oh SQLite is absolutely a reasonable inclusion and it's very helpful that it was
included in the higher order clog demos. It was only when I deployed that I
noticed I had to add a (trivial) foreign dependency, but it is absolutely not a
very big deal. Fair to say it's more my problem than clogs.

Yeah, having a clog core could be useful. Not very high priority, a
someday/maybe item, for I very much look forward to what you have in store
concerning data bindings.

On the topic of rucksack I've found [Nick Levine's](https://nicklevine.org/lisp-book/contents/chac2.pdf) (and the [previous chapter](https://nicklevine.org/lisp-book/contents/chac2.pdf)) and Edi
Weitz's [Common Lisp recipes in chapter 21-4](https://books.google.ca/books?id=FPxSCwAAQBAJ&printsec=frontcover&redir_esc=y#v=onepage&q&f=false) helpful. And this is a separate
idea, but I've also recently encountered [Kenny Tilton's cells](https://tilton.medium.com/the-cells-manifesto-b21ed10329f0) . An old GTK
version in [cellsGTK](https://cells-gtk.common-lisp.dev/). Mr. Tilton's more [recent iteration on the idea here](https://github.com/kennytilton/matrix). I
suppose the academic term is constraint based GUI programming. 
Not too clear on how this might work with presentations? I'm a beginner
myself so please take all of this with a grain of salt. I'm just brainstorming
and seeing what sticks on the wall; I hope it's of some use.

I did search for practical examples of pure lisp persistence solutions. Here's
what I was able to find. I clearly should be spending more time on my classes
than this, but here goes&#x2026;

I do know [the nyxt browser uses a common lisp in-memory datastore: cl-prevalence.](https://github.com/atlas-engineer/nyxt/blob/master/nyxt.asd) And
Rucksack's proprietary, distributed, more advanced cousin Allegrocache is used
by at least 2Is. Allegrocache was originally based on ObjectStore (Dan Weinreb
of Symbolics fame + others). [Dan does a fair job at defending object oriented
database management systems here.](https://web.archive.org/web/20110715000000*/https://www.danweinreb.org/blog/object-oriented-database-management-systems-succeeded) I would like to point out that [Glenn D. House
of 2Is (DoD contractor) testifies (21:30)](https://www.youtube.com/watch?v=MplVyKN0lHg) talks about his company's experience
with AllegroCache more here. Still, examples are pretty sparse. Pardon my french, 
but if I weren't down for some cowboy shit, I wouldn't have picked lisp. I use 
Rucksack [myself on my toy project (no intention of controversy!).](https://www.hanshenwang.com/bible?verses=1-1-1-1-1-31) 

If you happen do evaluate Rucksack yourself, I found that I really had to follow 
the advice (in doc/rucksack-tutorial.lisp) to manually open and close the store 
during the initial data migration, because I took a solid performance hit wrapping
 all invocations in a `with-rucksack` form. Though this is probably my own fault.
 I'm happy to say Rucksack and CLOG deploys on fly.io without a hitch!
You've mentioned teaching your little ones in an online CLOG demo and
fly.io has a great free tier.

I have been very impressed with the longevity of Common Lisp libraries, and
browsing both Rucksack and Cells mentioned above does feel like spelunking in
the ruins of the long lost civilization. I just can't seem to figure out if I'm
the barbarian and these libraries have been superseded by better methods, or if
there is nothing new under the sun and these artifacts have yet to be surpassed.
I suppose the Hebrews aren't entirely strangers to the idea of a dead language
coming back to life. Either way I'm very excited to see clog growing up!

Best,
